plugins/package-managers/maven/src/main/kotlin/Tycho.kt

```diff
@@ -329,3 +341,104 @@ internal class TychoProjectsCollector : AbstractMavenLifecycleParticipant() {
         logger.info { "Found ${builtProjects.size} projects during build." }
     }
 }
+
+private val tychoLogger = loggerOf(Tycho::class.java)
+
+/**
+ * Return a [PackageResolverFun] that can handle special OSGi artifacts detected during a Tycho analysis.
+ * OSGi bundles retrieved from a P2 repository cannot be handled by the default resolver function, which is not aware
+ * of the special repository layout. Also, they do not have a POM with metadata, but may contain some metadata in
+ * their OSGi manifest.
+ * This function first delegates to the given [delegate] function to handle normal packages in the usual way. If this
+ * fails, it checks whether the dependency can be found in the p2 cache of the local Maven repository pointed to by
+ * [localRepositoryRoot]. (Since the Tycho project has been built, the cache should be populated.) If this is the case,
+ * it tries to obtain metadata from the OSGi manifest of the bundle. This approach is partly inspired by
+ * https://github.com/OpenNTF/p2-layout-provider.
+ */
+internal fun tychoPackageResolverFun(
+    delegate: PackageResolverFun,
+    localRepositoryRoot: File = RepositorySystem.defaultUserLocalRepository
+): PackageResolverFun =
+    { dependency ->
+        runCatching {
+            delegate(dependency)
+        }.recoverCatching { exception ->
+            createPackageFromLocalArtifact(dependency.artifact, localRepositoryRoot)
+                ?: throw exception
+        }.getOrThrow()
+    }
+
+/**
+ * Create a [Package] for the given [artifact] obtaining metadata from the given [manifest]. This function is used for
+ * OSGi artifacts for which no POM is available.
+ */
+internal fun createPackageFromManifest(artifact: Artifact, manifest: Manifest): Package =
+    with(manifest.mainAttributes) {
+        val declaredLicenses = setOfNotNull(getValue("Bundle-License"))
+        val declaredLicensesProcessed = processDeclaredLicenses(declaredLicenses)
+        val authors = setOfNotNull(getValue("Bundle-Vendor"))
+
+        val vcs = parseScm(getValue("Eclipse-SourceReferences").toScm(), artifact.identifier())
+
+        Package(
+            id = Identifier(
+                type = "Maven",
+                namespace = artifact.groupId,
+                name = artifact.artifactId,
+                version = artifact.version
+            ),
+            declaredLicenses = declaredLicenses,
+            declaredLicensesProcessed = declaredLicensesProcessed,
+            description = getValue("Bundle-Description").orEmpty(),
+            homepageUrl = "",
+            binaryArtifact = RemoteArtifact.EMPTY,
+            sourceArtifact = RemoteArtifact.EMPTY,
+            vcs = vcs,
+            vcsProcessed = vcs,
+            authors = authors
```

Comment:
Nit: This could precede `declaredLicenses` to match the property declaration order.